### PR TITLE
Alerting: Limit and clean up old alert rules versions

### LINKED ISSUE #73856: Alerting: Evict alert history over time
### What happened?

The `alert_configuration` table has a system to evict history as it gets very old - but the `alert_rule_version` table does not.

This can result in the rule history table growing unboundedly, one user reported a 13gb table over a very long period of use.

We should use a similar eviction system for that table.

### What did you expect to happen?

I expected the rule version history table to be bounded like the config history one is.

### Did this work before?

No

### How do we reproduce it?

1. Use sqlite to make things easy
2. Update a rule repeatedly - hundreds of times
3. Watch the database size grow forever

### Is the bug inside a dashboard panel?

_No response_

### Environment (with versions)?

Grafana: main
OS:
Browser:

### Grafana platform?

None

### Datasource(s)?

_No response_

### PR DESCRIPTION
**Pull request by igloo12**
This is a revived version of this closed PR with all the comments addressed https://github.com/grafana/grafana/pull/76084
@JacobsonMT @yuri-tceretian

What is this feature?

This PR deletes alert_rule_version records for an alert rule in an org having excess versions stored than the default limit. The default limit RuleVersionRecordLimits has been set to 100. That means 100 versions are stored for a particular alert rule in an org.

Why do we need this feature?

So that the rule history table doesn't grow unboundedly.

Who is this feature for?

All users having significant number of alert rule updates over time.

Which issue(s) does this PR fix?:

Fixes https://github.com/grafana/grafana/issues/73856

Special notes for your reviewer:

Please check that:

 It works as expected from a user's perspective.
 If this is a pre-GA feature, it is behind a feature toggle.
 The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/ngalert/store/alert_rule.go
Comment (JacobsonMT) [inline pkg/services/ngalert/store/alert_rule.go:289]: NIT: Looks like we've split the receivers between pointer and value on DBstore, not sure if it was intentional or not, but it's probably worth staying consistent within the file for now.
```suggestion
func (st DBstore) deleteOldAlertRuleVersions(ctx context.Context, ruleUID string, orgID int64, limit int) (int64, error) {
```

File: pkg/services/ngalert/store/alert_rule.go
Comment (JacobsonMT) [inline pkg/services/ngalert/store/alert_rule.go:40]: 100 is a fairly large amount of alert rule versions and seems like a good conservative number. That being said, I still wonder if this should be made opt-in for now. Preferably by making the limit configurable and default to 0 (unlimited).

Most will likely see the cleanup as a plus but I can imagine it being unexpected to some users who might want to keep more version history to suddenly have them deleted in a minor version update. We can always make it opt-out later.

I can point you to some example code to work from or I can add the configuration code myself if you prefer not to.

File: pkg/services/ngalert/store/alert_rule.go
Comment (JacobsonMT) [inline pkg/services/ngalert/store/alert_rule.go:317]: Related to your comment about off-by-one, you're right. This should either be `id < ?` or the threshold should be `highest.ID - 1`

File: pkg/services/ngalert/store/alert_rule_test.go
Comment (JacobsonMT) [inline pkg/services/ngalert/store/alert_rule_test.go:1176]: NIT: This call is not necessary since `deleteOldAlertRuleVersions` should be called from `UpdateAlertRules`.

File: pkg/services/ngalert/store/alert_rule_test.go
Comment (JacobsonMT) [inline pkg/services/ngalert/store/alert_rule_test.go:1228]: NIT: This call is only necessary because of the hardcoded limit of 100. If the value were available as a config, you would not need to call a private method and rely on the calls to `store.UpdateAlertRules` to do the cleanup.

**Review by JacobsonMT — CHANGES_REQUESTED** [review]
Thanks for picking this up from the previous PR! This is a good feature and the implementation just needs a couple adjustments before I can approve. :rocket:

File: go.work.sum
Comment (JacobsonMT) [inline go.work.sum:1]: Feel free to leave out changes to this file in this PR instead of having to fix conflicts.

**Comment by igloo12** [comment]
Bad rebase causes everyone to be tagged in the PR review  :(  Sorry fixing

File: go.work.sum
Comment (igloo12) [reply go.work.sum:1]: done

File: pkg/services/ngalert/store/alert_rule_test.go
Comment (igloo12) [reply pkg/services/ngalert/store/alert_rule_test.go:1235]: Done

File: pkg/services/ngalert/store/alert_rule_test.go
Comment (igloo12) [reply pkg/services/ngalert/store/alert_rule_test.go]: Done

**Comment by igloo12** [comment]
@JacobsonMT I fixed the rebase issue but can't change the reviewer or labels. Do you want me to open another PR or can you update this one?

File: pkg/services/ngalert/store/alert_rule.go
Comment (igloo12) [reply pkg/services/ngalert/store/alert_rule.go:312]: The issue is id < ? is causing it to be 1 less than the threshold. Should I update the other method to match id <= ? in this PR?

File: pkg/services/ngalert/store/alert_rule.go
Comment (igloo12) [reply pkg/services/ngalert/store/alert_rule.go]: I took a stab at adding it to the configuration. Let me know if there is a better place to add the configuration.

File: pkg/services/ngalert/store/alert_rule.go
Comment (igloo12) [reply pkg/services/ngalert/store/alert_rule.go:312]: This should be id <= ? becuase I am grabbing the first item that needs to be deleted

**Comment by yuri-tceretian** [comment]
No worries. I cleaned up labels and reviewers

**Comment by igloo12** [comment]
Thanks @yuri-tceretian. The build is getting blocked on each build with the message `Build is blocked, please, contact repo admin in order to proceed` Can you please unblock it? Thanks

**Comment by yuri-tceretian** [comment]
The test failed 
```
FAIL: TestIntegration_AlertRuleVersionsCleanup (0.02s)

    --- FAIL: TestIntegration_AlertRuleVersionsCleanup/when_calling_the_cleanup_with_fewer_records_than_the_limit_all_records_should_stay (0.00s)

        alert_rule_test.go:1312: 

            	Error Trace:	/drone/src/pkg/services/ngalert/store/alert_rule_test.go:1312

            	Error:      	Received unexpected error:

            	            	sql: Scan error on column index 0, name "id": unsupported Scan, storing driver.Value type int64 into type *models.AlertRuleVersion

            	Test:       	TestIntegration_AlertRuleVersionsCleanup/when_calling_the_cleanup_with_fewer_records_than_the_limit_all_records_should_stay

    --- FAIL: TestIntegration_AlertRuleVersionsCleanup/only_oldest_records_surpassing_the_limit_should_be_deleted (0.00s)

        alert_rule_test.go:1364: 

            	Error Trace:	/drone/src/pkg/services/ngalert/store/alert_rule_test.go:1364

            	Error:      	Received unexpected error:

            	            	sql: Scan error on column index 0, name "id": unsupported Scan, storing driver.Value type int64 into type *models.AlertRuleVersion

            	Test:       	TestIntegration_AlertRuleVersionsCleanup/only_oldest_records_surpassing_the_limit_should_be_deleted

FAIL
```

**Comment by igloo12** [comment]
@yuri-tceretian  Thanks, I updated the DB lookup to use Find instead of Get

**Comment by igloo12** [comment]
@yuri-tceretian can you please unblock the integration tests when you can? Thanks

File: pkg/setting/setting.go
Comment (yuri-tceretian) [inline pkg/setting/setting.go:514]: These are legacy settings. We can drop it from here

Comment (igloo12) [reply pkg/setting/setting.go:514]: @yuri-tceretian This was my best guess. Can you point to the best place to add these settings? Thanks

Comment (igloo12) [reply pkg/setting/setting.go]: @yuri-tceretian thanks updated

**Comment by igloo12** [comment]
@yuri-tceretian can you please release the block on the tests for this build? Thanks

**Comment by igloo12** [comment]
@yuri-tceretian A friendly bump to allow the unit tests to run. Thanks

**Comment by yuri-tceretian** [comment]
sorry! I thought I approved the build. Done

**Comment by yuri-tceretian** [comment]
The branch was pretty old and CI failed. I merged the master branch into your branch and updated some code.

**Comment by yuri-tceretian** [comment]
Just food for thought: what if we allow 0, and basically let users avoid saving data into version table?

**Comment by igloo12** [comment]
@yuri-tceretian It looks like the build is failing with a known flaky test https://drone.grafana.net/grafana/grafana/197448

**Comment by igloo12** [comment]
@yuri-tceretian I updated the branch with  https://github.com/grafana/grafana/pull/93491. Can you please run the tests on the PR? Thanks

**Comment by yaskinny** [comment]
This bug is really frustrating. Has anyone reviewed this PR? Will it be merged?

@JacobsonMT @yuri-tceretian

File: conf/defaults.ini
Comment (yuri-tceretian) [inline conf/defaults.ini:1345]:
```suggestion
rule_version_record_limit = 0
```

File: conf/sample.ini
Comment (yuri-tceretian) [inline conf/sample.ini:1334]:
```suggestion
;rule_version_record_limit= 0
```

File: conf/defaults.ini
Comment (yuri-tceretian) [inline conf/defaults.ini:1344]:
```suggestion
# Defines the limit of how many alert rule versions
# should be stored in the database for each alert rule in an organization including the current one.
# 0 value means no limit
```

File: pkg/setting/setting_unified_alerting.go
Comment (yuri-tceretian) [inline pkg/setting/setting_unified_alerting.go:463]:
```suggestion
	uaCfg.RuleVersionRecordLimit = ua.Key("rule_version_record_limit").MustInt(0)
	if uaCfg.RuleVersionRecordLimit < 0 {
	     return fmt.Errorf("setting 'rule_version_record_limit' is invalid, only 0 or a positive integer are allowed")
	}
```

**Review by yuri-tceretian — APPROVED** [review]
LGTM. Please address my comments and we can merge

**Comment by igloo12** [comment]
@yuri-tceretian Thanks for the review. I added your suggestions. You will need to run the tests.

**Review by JacobsonMT — APPROVED** [review]
LGTM :rocket: Great work!

**Pull request merged**
